src/main/java/org/apache/spark/shuffle/sort/RssUnsafeShuffleWriter.java

```diff
@@ -0,0 +1,248 @@
+        this.blockManager = blockManager;
+        this.memoryManager = memoryManager;
+        this.serializer = dependency.serializer().newInstance();
+        this.partitioner = dependency.partitioner();
+        this.writeMetrics = taskContext.taskMetrics().shuffleWriteMetrics();
+        this.taskContext = taskContext;
+        this.sparkConf = sparkConf;
+        this.initialSortBufferSize = sparkConf.getInt("spark.shuffle.sort.initialBufferSize",
+                DEFAULT_INITIAL_SORT_BUFFER_SIZE);
+        open();
+    }
+
+    private void updatePeakMemoryUsed() {
+        // sorter can be null if this writer is closed
+        if (sorter != null) {
+            long mem = sorter.getPeakMemoryUsedBytes();
+            if (mem > peakMemoryUsedBytes) {
+                peakMemoryUsedBytes = mem;
+            }
+        }
+    }
+
+    /**
+     * Return the peak memory used so far, in bytes.
+     */
+    public long getPeakMemoryUsedBytes() {
+        updatePeakMemoryUsed();
+        return peakMemoryUsedBytes;
+    }
+
+    @Override
+    public int recordsWritten() {
+        return recordsWritten;
+    }
+
+    @Override
+    public double reductionFactor() {
+        return 0;
+    }
+
+    @Override
+    public int collectionSizeInBytes() {
+        return 0;
+    }
+
+    @Override
+    public ShuffleWriteTimeMetadata getShuffleWriteTimeMetadata() {
+        return metadata._2;
+    }
+
+    @Override
+    public ShuffleWriteMetadata getShuffleWriteMetadata() {
+        return metadata._1;
+    }
+
+    @Override
+    public void addRecord(int partitionId, Product2<K, V> record) {
+        // Keep track of success so we know if we encountered an exception
+        // We do this rather than a standard try/catch/re-throw to handle
+        // generic throwables.
+        // TODO: recordsWritten fix this
+        recordsWritten += 1;
+        try {
+            insertRecordIntoSorter(record);
+        } catch (Throwable e) {
+            logger.error("Error while adding a new record", e);
+            if (sorter != null) {
+                sorter.cleanupResources();
+            }
+            throw new RssException(e.getMessage());
+        }
+    }
+
+    private void open() {
+        assert (sorter == null);
+        sorter = new RssShuffleExternalSorter(
+                memoryManager,
+                blockManager,
+                taskContext,
+                initialSortBufferSize,
+                partitioner.numPartitions(),
+                sparkConf,
+                writeMetrics,
+                this);
+        serBuffer = new MyByteArrayOutputStream(DEFAULT_INITIAL_SER_BUFFER_SIZE);
+        serOutputStream = serializer.serializeStream(serBuffer);
+    }
+
+    @VisibleForTesting
+    void insertRecordIntoSorter(Product2<K, V> record) throws IOException {
+        assert(sorter != null);
+        final K key = record._1();
+        final int partitionId = partitioner.getPartition(key);
+        serBuffer.reset();
```

Comment:
I remember some serializer like JavaSerializer may use object reference in serialized stream, each serialized object inside the serialized stream will not be independent. Thus getting the bytes from MyByteArrayOutputStream each time after adding key/value may not be always safe.

KyroSerializer does not use object reference (if I remember correctly), and will be safe here.

Maybe add some check here to make sure the serializer is KyroSerializer?

Reply:
We should always use kyroSerializer, can we just fail if we are not using Kyro